File: data_structures/linked_list/typescript/SinglyLinkedListNode.ts
import { ILinkedListNode } from '@ds/linked_list/typescript/types';

export default class SinglyListNode<T> implements ILinkedListNode<T> {
    val: T;
    next: SinglyListNode<T> | null = null;
    key: number | string;

    constructor(val: T = undefined, key: number | string = null, next: SinglyListNode<T> | null = null) {
        this.val = val;
        this.next = next;
        this.key = key;
    }

    /**
     * Create a sningly linked list from array
     * @param vals: array of values
     */
    fromArray(vals: T[]): this {
        let counter = 0;
        if (this.val === undefined) {
            this.val = vals[0];
            counter = 1;
        }

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let current: SinglyListNode<T> | null = this;

        for (let i = counter; i < vals.length; i++) {
            const node = new SinglyListNode<T>(vals[i], null);
            current.next = node;
            current = node;
        }

        return this;
    }

    /**
     * Convert singly linked list to array
     */
    toArray(): T[] {
        const result: T[] = [];

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let current: ILinkedListNode<T> | null = this;

        while (current) {
            result.push(current.val);
            current = current.next;
        }

        return result;
    }

    /**
     * Reverse singly linked list and return new head
     */
    reverse(): SinglyListNode<T> {
        if (!this.val) {
            return this;
        }

        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let current: SinglyListNode<T> | null = this;
        let next: SinglyListNode<T> | null = null;
        let prev: SinglyListNode<T> | null = null;

        while (current) {
            next = current.next;
            current.next = prev;

            prev = current;
            current = next;
        }

        return prev;
    }
}

File: data_structures/linked_list/python/singly_linked_list_node.py
from typing import Union, List, TypeVar, Generic

T = TypeVar("T")


class ListNode(Generic[T]):
    key: int
    next: Union[None, 'ListNode'] = None

    def __init__(self, val: T = None, key: int = None):
        self.val = val
        self.key = key

    def __repr__(self):
        return f"ListNode({self.val})"

    def from_array(self, vals: List[T]) -> 'ListNode[T]':
        """
        Create a singly linked list from array
        """
        counter = 0

        if len(vals) <= counter:
            pass

        if self.val is None:
            self.val = vals[0]
            counter += 1

        current = self

        for i in range(counter, len(vals)):
            node = ListNode(vals[i])
            current.next = node
            current = node

        return self

    def to_array(self, to_string=False) -> List[T]:
        """
        Convert singly linked list to array
        """
        current = self
        result: List[T] = []

        while current:
            result.append(str(current.val) if to_string else current.val)
            current = current.next

        return result

    def reverse(self) -> 'ListNode[T]':
        """
        Reverse singly linked list
        """
        current_node = self
        prev_node = None
        next_node = None

        while current_node:
            next_node = current_node.next
            current_node.next = prev_node

            prev_node = current_node
            current_node = next_node

        return prev_node
